File: 1143.longest-common-subsequence/solution.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        M, N = len(text1), len(text2)
        dp = [[0] * (N + 1) for _ in range(M + 1)]
        
        for i in range(1, M + 1):
            for j in range(1, N + 1):
                # Pick larger counts from the two previous statuses.
                dp[i][j] = max(
                    dp[i - 1][j],
                    dp[i][j - 1]
                )
                
                if text1[i - 1] != text2[j - 1]:
                    continue
                
                dp[i][j] = max(
                    # do nothing
                    dp[i][j],
                    
                    # if matched, increment the count from the previously matched count
                    dp[i - 1][j - 1] + 1
                )
        
        # for i in range(M + 1):
        #     print(dp[i])
        return dp[M][N]

File: 1143.longest-common-subsequence/solution.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int len1 = text1.length();
        int len2 = text2.length();
        
        int[][] dp = new int[len1 + 1][len2 + 1];
        for (int i = 0; i < len1; i++) {
            for (int j = 0; j < len2; j++) {
                dp[i + 1][j + 1] = Math.max(
                    dp[i][j + 1],
                    dp[i + 1][j]
                );
                if (text1.charAt(i) != text2.charAt(j)) {
                    continue;
                }
                dp[i + 1][j + 1] = Math.max(
                    dp[i + 1][j + 1],
                    dp[i][j] + 1
                );
            }
        }
        
        return dp[len1][len2];
    }
}